source/aws_imds_client.c

```diff
@@ -460,11 +587,28 @@ static int s_make_imds_http_query(
 }
 
 typedef void(imds_state_fn)(struct aws_imds_client_user_data *);
+static void s_query_resource(struct aws_imds_client_user_data *user_data);
 
 /* Make an http request to put a ttl and hopefully get a token back. */
 static void s_query_token(struct aws_imds_client_user_data *user_data) {
     AWS_PRECONDITION(user_data->query_state == AWS_IMDS_QS_TOKEN_REQ);
 
+    /**
+     * If this requester shouldn't update token, meaning either the token is available,
+     * or the token is updating by other requester, this requester should just either copy
+     * token or wait for the update then copy.
+     */
+    while (!s_requester_test_and_try_dominate_cached_token_update(user_data)) {
```

Comment:
Is this spinning on a pending token query?  If so, we'll need to rework this approach substantially.  The cached credentials provider faces a similar situation: arbitrary incoming async get requests but the "real" singleton query could be outstanding or need to be made.  It solves it by actually tracking all the requests and completing them when the query completes either successfully or unsuccessfully.  Meanwhile those get requests return immediately; we can't block or spin anywhere.